### PR TITLE
Return IChannelSummarizeResult from SharedObject's summarize

### LINKED ISSUE #4402: Update summarize to also return a list of references to fluid objects.
Summarize should also return a list of references to fluid objects. This will be used to form the "reference graph" that garbage collection will run on.

### PR DESCRIPTION
**Pull request by agarwal-navin**
Updated summarize() API in SharedObject to return IChannelSummarizeResult which includes garbage collection nodes.

Added unit tests in DDS to test it generates GC nodes correctly.

Fixes #4402, #4127

### CODE REVIEW

File: packages/dds/map/src/test/directory.spec.ts [hunk 1]
```diff
@@ -1036,5 +1036,147 @@ describe("Directory", () => {
...
+                containerRuntimeFactory.processAllMessages();
+
+                // Verify the GC nodes returned by summarize.
+                const gcNodes = directory2.summarize().gcNodes;
+                assert.strictEqual(gcNodes.length, 1, "There should only be one GC node in summary");
+                assert.strictEqual(gcNodes[0].id, "/", "GC node's id should be /");
+                assert.deepStrictEqual(
+                    gcNodes[0].outboundRoutes,
+                    [
+                        subMap.handle.absolutePath,
```
Comment (vladsud) [inline packages/dds/map/src/test/directory.spec.ts:1066]: Nit: order is not important here (from API perspective), but is important for test to pass.
We have same problem in summarization tests. Should we do something about it? :)

File: packages/dds/map/src/test/map.spec.ts [hunk 2]
```diff
@@ -456,5 +456,113 @@ describe("Map", () => {
...
+                const gcNodes = map2.summarize().gcNodes;
+                assert.strictEqual(gcNodes.length, 1, "There should only be one GC node in summary");
+                assert.strictEqual(gcNodes[0].id, "/", "GC node's id should be /");
+                assert.deepStrictEqual(
+                    gcNodes[0].outboundRoutes,
+                    [subMap.handle.absolutePath],
+                    "GC node's outbound routes is incorrect");
+            });
+
+            it("can generate GC nodes when handles are removed from data", () => {
```
Comment (vladsud) [inline packages/dds/map/src/test/map.spec.ts:477]: Is there any way to reuse logic across different DDSs?
It feels like you can build an interface that does various common things (like add one reference, add multiple reference to same thing, add 2 references, summarize) and then have each DDS test code implement such interface and have "core" / "test driver" to be shared across all of them somewhere in shared test code.
Would it be useful?

**Review by vladsud — APPROVED** [review]

File: packages/dds/map/src/test/directory.spec.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/dds/map/src/test/directory.spec.ts:1066]: You right! We might end up in similar problems. We should definitely do something about it. I will fix this :-).

File: packages/dds/map/src/test/map.spec.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/dds/map/src/test/map.spec.ts:477]: I think it might be useful. It would be convenient to add more tests that apply to all these DDSs.
I thought of something similar but didn't do it because I wasn't sure if wanted to share test code across DDS (There is no such code today and sharing is limited to a DDS package).
But it makes sense to share because its a lot of duplicate code. I will probably give it a try and see how we feel about them.

Comment (anthony-murphy) [reply packages/dds/map/src/test/map.spec.ts:477]: @markfields just is adding some shared test code in #4422. could be a good landing place

Comment (agarwal-navin) [reply packages/dds/map/src/test/map.spec.ts:477]: I started adding it to a "fluid-internal/test-dds-utils" package under "packages/dds". I don't want to leak it outside the dds package.

Comment (markfields) [reply packages/dds/map/src/test/map.spec.ts:477]: I'd lean towards `test-runtime-utils` rather than a new package. This would be a fine fit there, I think - this package already has a dependency on it.

**Comment by agarwal-navin** [comment]
@vladsud @anthony-murphy Moved the GC tests a new fluid-internal test util under packages/dds. Can you please take a look and let me know if this looks fine.

File: packages/dds/map/src/test/map.spec.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/dds/map/src/test/map.spec.ts]: The utils that I am adding won't be used outside of the packages/dds directory. Also, test-runtime-utils is not an internal package and I would not prefer to add utilities to it that won't be used externally.

Comment (agarwal-navin) [reply packages/dds/map/src/test/map.spec.ts]: However, if we decide that these utils should be used outside the package and / or it should be exposed externally, it can be moved to test-runtime-utils.

File: packages/dds/map/src/test/directory.spec.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/dds/map/src/test/directory.spec.ts]: Sorted the arrays before comparing.

**Comment by markfields** [comment]
@agarwal-navin the new package should be listed in the Test layer of the layer info JSON file, for layerCheck. And it needs a Dev flag somewhere to ensure no prod code takes a dependency on it.

File: packages/dds/cell/src/test/cell.spec.ts [hunk 3]
```diff
@@ -132,29 +133,83 @@ describe("Cell", () => {
...
+                    nestedObj: {
+                        subcell2Handle: newSubCell2.handle,
+                    },
+                };
+                cell.set(containingObject);
+                expectedRoutes = [ newSubCell.handle.absolutePath, newSubCell2.handle.absolutePath ];
+                containerRuntimeFactory.processAllMessages();
+            }
+
+            const gcTestProvider: IGCTestProvider = {
```
Comment (vladsud) [inline packages/dds/cell/src/test/cell.spec.ts:204]: If you squeeze, you can see that all the code under describe() is basically a class (I'll call it GCCellTestProvider), followed by one-liner - runGCTests(new GCCellTestProvider()).
There is no need for beforeEach(), local variables (these are member variables of class).
IMHO class better represents the intention and relationships between methods and variables, otherwise one continuously needs to ask questions like "what's the scope/ lifetime of this variable?"

File: packages/dds/test-dds-utils/src/gcTestRunner.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+import { strict as assert } from "assert";
+import { ISharedObject } from "@fluidframework/shared-object-base";
+
+/**
+ * Defines a set of functions to be passed to the GC test runner.
+ */
+export interface IGCTestProvider {
+    /** Returns the DDS whose GC data is to be verified */
+    getSharedObject(): ISharedObject;
```
Comment (vladsud) [inline packages/dds/test-dds-utils/src/gcTestRunner.ts:14]: we usually implement it as readonly property - useful to have it that way for consistency with rest of code

**Comment by agarwal-navin** [comment]
> @agarwal-navin the new package should be listed in the Test layer of the layer info JSON file, for layerCheck. And it needs a Dev flag somewhere to ensure no prod code takes a dependency on it.

Thanks Mark! Added to layerInfo.json

File: packages/dds/cell/src/test/cell.spec.ts (hunk 3 above)
Comment (agarwal-navin) [reply packages/dds/cell/src/test/cell.spec.ts]: Great suggestion! I changed it to a class.
I put the class definition in the describe block itself for each test. Should it be outside the describe blocks or perhaps in a separate file?

File: packages/dds/cell/src/test/cell.spec.ts [hunk 5]
```diff
@@ -161,55 +161,55 @@ describe("Cell", () => {
...
-            async function deleteOutboundRoutes() {
-                cell.delete();
-                expectedRoutes = [];
-                containerRuntimeFactory.processAllMessages();
+            class GCSharedCellProvider implements IGCTestProvider {
+                private subCellCount = 0;
+                private _expectedRoutes: string[] = [];
+
+                constructor() {
+                    this.subCellCount = 0;
```
Comment (vladsud) [inline packages/dds/cell/src/test/cell.spec.ts:169]: constructor is not needed - you already init variables where they are declared

File: packages/dds/cell/src/test/cell.spec.ts [hunk 6]
```diff
@@ -161,55 +161,55 @@ describe("Cell", () => {
...
+                    this._expectedRoutes = [];
+                }
+
+                public get sharedObject() {
+                    // Return the remote SharedCell because we want to verify its summary data.
+                    return cell2;
```
Comment (vladsud) [inline packages/dds/cell/src/test/cell.spec.ts:175]: Are you sure we want to reuse same object across multiple test instances? I.e. shouldn't it be member of this class, i.e. created and initialized every time instance of this class is created?

File: packages/dds/cell/src/test/cell.spec.ts (hunk 3 above)
Comment (vladsud) [reply packages/dds/cell/src/test/cell.spec.ts]: I've looked at just one implementation, and asked question about cell2 lifetime management. I think if the intention for this class to be self-contained, then it's better to move it out to global scope, as this will ensure compiler will help you find dependencies on state that should not be there. But if the intention is to use some of the state setup above, then you can either leave it in place, or find other ways to pass it (like creating a factory function that returns new instance of class bound to some state, similar how mixins do it - see for example mixinRequestHandler.

Basically let's try to apply same quality to test code - how would you write similar shipping code to ensure that it's easier to find bugs and understand scoping / lifetime management of various pieces?
I find that our test code is quite often harder to read then shipping code :(

Otherwise looks much better IMHO!

Comment (agarwal-navin) [reply packages/dds/cell/src/test/cell.spec.ts]: I thought about this too. I wanted to re-use the setup above. But I think I can move the creation of the DDS to a function and call it from the constuctor.
I agree about our test code not being readable! I will try to clean it up as much as I can here :-).

File: packages/dds/cell/src/test/cell.spec.ts (hunk 5 above)
Comment (agarwal-navin) [reply packages/dds/cell/src/test/cell.spec.ts:169]: oh yeah, this is class now!

File: packages/dds/cell/src/test/cell.spec.ts (hunk 6 above)
Comment (agarwal-navin) [reply packages/dds/cell/src/test/cell.spec.ts:175]: Yeah, you are right. I will update it.

File: packages/dds/cell/src/test/cell.spec.ts (hunk 3 above)
Comment (agarwal-navin) [reply packages/dds/cell/src/test/cell.spec.ts]: Okay. So I made the class independent. Moved out creation of DDSs into a separate function which is called by the class and from the rest of the test code.
I cleaned up the tests as much as I could for now. I think they can be cleaned further but I would rather do that in a separate change.

File: packages/dds/cell/src/test/cell.spec.ts (hunk 6 above)
Comment (agarwal-navin) [reply packages/dds/cell/src/test/cell.spec.ts]: Updated this as per my other comment.
